src/Umbraco.Infrastructure/Scoping/CallContext.cs

```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Umbraco.Core.Scoping
+{
+    /// <summary>
+    /// Provides a way to set contextual data that flows with the call and
+    /// async context of a test or invocation.
+    /// </summary>
+    public static class CallContext
+    {
+        private static readonly ConcurrentDictionary<string, Guid?> _state = new ConcurrentDictionary<string, Guid?>();
+
+        /// <summary>
+        /// Stores a given object and associates it with the specified name.
+        /// </summary>
+        /// <param name="name">The name with which to associate the new item in the call context.</param>
+        /// <param name="data">The object to store in the call context.</param>
+        public static void SetData(string name, Guid? data)
+        {
+            _state[name + Thread.CurrentThread.ManagedThreadId] = data;
+        }
+
+
+        /// <summary>
+        /// Retrieves an object with the specified name from the <see cref="CallContext"/>.
+        /// </summary>
+        /// <param name="name">The name of the item in the call context.</param>
+        /// <returns>The object in the call context associated with the specified name, or <see langword="null"/> if not found.</returns>
+        public static Guid? GetData(string name)
+        {
+            return _state.TryGetValue(name + Thread.CurrentThread.ManagedThreadId, out var data) ? data : null;
```

Comment:
This isn't going to work as expected, this isn't going to flow at all with async operations and will only work for non-async/singlethreaded operations which means it won't work for any async operations like webapi calls so our Scopes will not be tracked properly and we'll leak transactions, etc... . 

The CallContext.LogicalGetData/LogicalSetData which is what we were using **does** flow with async. Here's the info https://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html this link is referenced in our codebase in `HybridAccessorBase`. 

IIRC AsyncLocal is the replacement for CallContext.LogicalSetData/LogicalGetData. I know you've seen a few topics about this such as https://stackoverflow.com/questions/31707362/how-do-the-semantics-of-asynclocal-differ-from-the-logical-call-context and I know you've seen this one too http://www.cazzulino.com/callcontext-netstandard-netcore.html (also note the strongly typed example at the bottom which doesn't require boxing/unboxing).

We will need to use AsyncLocal in place of CallContext since we must have the data flow with async behaviors else our scopes/transactions will not work.

Reply:
I tried using implementation like here http://www.cazzulino.com/callcontext-netstandard-netcore.html but that did not pass our tests..

Reply:
Cool, i'll play around with this today and report back.